File: src/actions_py/actions_py/move_to_position_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from my_robot_interfaces.action import MoveToPos
from my_robot_interfaces.msg import Trigger


class MoveToPosClientNode(Node):
    def __init__(self):
        super().__init__("move_to_pos_client")
        self.goal_handle_ = None
        self.move_to_pos_client_ = ActionClient(self, MoveToPos, "move_to_pos")
        self.cancel_goal_sub_ = self.create_subscription(Trigger, "cancel_move_to_pos_goal", self.callback_cancel_move,10)

        self.get_logger().info("move_to_pos_client has been started")

    def send_goal(self, target_position, velocity):
        #wait for the server
        while not self.move_to_pos_client_.wait_for_server(timeout_sec=1.0):
            self.get_logger().warn("Waiting for Server....")
            
        #creat a goal
        
        goal = MoveToPos.Goal()
        goal.target_position = target_position
        goal.velocity = velocity
        
        #sending the gaol
        
        self.get_logger().info("sending the goal")
        self.move_to_pos_client_.send_goal_async(goal, feedback_callback=self.goal_feedback_callback).\
                                add_done_callback(self.goal_resopnse_callback)
                                

    def callback_cancel_move(self, msg):
        self.cancel_goal()
        

    def cancel_goal(self):
        if self.goal_handle_ is not None:
            self.get_logger().info("send a cancel request")
            self.goal_handle_.cancel_goal_async()

        
        
        
    def goal_resopnse_callback(self, future):
        self.goal_handle_: ClientGoalHandle= future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal got accepted")
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)
        else:
            self.get_logger().warn("Goal got rejected")
            
    def goal_result_callback(self,future):
        status = future.result().status
        result = future.result().result
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Success")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn("Canceled")
        self.get_logger().info("result:" + "Position: " + str(result.position) + " " + "Massage: " + result.message)
        
    def goal_feedback_callback(self, feedback_msg):
        position = feedback_msg.feedback.current_position
        self.get_logger().info("got feedback: " + str(position))
        


def main(args=None):
    rclpy.init(args=args)
    node = MoveToPosClientNode()
    node.send_goal(80, 1)
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/actions_cpp/src/move_to_position_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_to_pos.hpp"
#include "my_robot_interfaces/msg/trigger.hpp"

using MoveToPos = my_robot_interfaces::action::MoveToPos;
using MoveToPosGoalHandle = rclcpp_action::ClientGoalHandle<MoveToPos>;
using namespace std::placeholders;

class MoveToPosClientNode : public rclcpp::Node
{
public:
    MoveToPosClientNode() : Node("move_to_pos_client")
    {
        move_to_pos_client_ = rclcpp_action::create_client<MoveToPos>(this, "move_to_pos");

        cancel_sub_ = this->create_subscription<my_robot_interfaces::msg::Trigger>("cancel_move_to_pos_goal", 10, std::bind(&MoveToPosClientNode::cancel_goal, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "move_to_pos_client has been started");
    }

    void send_goal(int target_position, int velocity)
    {
        // wait for the action server
        move_to_pos_client_->wait_for_action_server();

        // create gaol
        auto goal = MoveToPos::Goal();
        goal.target_position = target_position;
        goal.velocity = velocity;

        // Add callbacks
        auto options = rclcpp_action::Client<MoveToPos>::SendGoalOptions();
        options.result_callback = std::bind(&MoveToPosClientNode::get_result_callback, this, _1);
        options.goal_response_callback = std::bind(&MoveToPosClientNode::goal_response_callback, this, _1);
        options.feedback_callback = std::bind(&MoveToPosClientNode::goal_feedback_callback, this, _1, _2);

        // Send the goal
        RCLCPP_INFO(this->get_logger(), "Sending a goal");
        move_to_pos_client_->async_send_goal(goal, options);

    }

    void cancel_goal(const my_robot_interfaces::msg::Trigger msg)
    {
        (void)msg;
        cancel_callback();
    }

private:
    void cancel_callback()
    {
        if (this->goal_handle_)
        {

            RCLCPP_INFO(this->get_logger(), "Cancel the goal");
            move_to_pos_client_->async_cancel_goal(goal_handle_);
            goal_handle_.reset();
            
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "goal cancle requested but there is no goal to cancel");
            
        }
        
    }

    // callback to know if the gaol is accepted or rejected
    void goal_response_callback(const MoveToPosGoalHandle::SharedPtr &goal_handle)
    {
        if (!goal_handle)
        {
            // rejected
            RCLCPP_INFO(this->get_logger(), "goal got rejected");
        }
        else
        {
            // accepted
            this->goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(), "goal got accepted");
        }
    }

    // callback ro recieve feedback during goal execution
    void goal_feedback_callback(const MoveToPosGoalHandle::SharedPtr &goal_handle, const std::shared_ptr<const MoveToPos::Feedback> feedback)
    {
        (void)goal_handle;
        int current_position = feedback->current_position;
        RCLCPP_INFO(this->get_logger(), "Got feedback: %d", current_position);
    }

    // calback to recive the result once the goal is done
    void get_result_callback(const MoveToPosGoalHandle::WrappedResult &result)
    {
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED)
        {
            RCLCPP_INFO(this->get_logger(), "Succeeded");
        }
        else if (status == rclcpp_action::ResultCode::ABORTED)
        {
            RCLCPP_ERROR(this->get_logger(), "Aborted");
        }
        else if (status == rclcpp_action::ResultCode::CANCELED)
        {
            RCLCPP_WARN(this->get_logger(), "Canceled");
        }

        int position = result.result->position;
        std::string message = result.result->message;
        RCLCPP_INFO(this->get_logger(), "Result: position: %d, message: %s", position, message.c_str());
    }

    rclcpp_action::Client<MoveToPos>::SharedPtr move_to_pos_client_;
    rclcpp::TimerBase::SharedPtr timer_;
    MoveToPosGoalHandle::SharedPtr goal_handle_;
    rclcpp::Subscription<my_robot_interfaces::msg::Trigger>::SharedPtr cancel_sub_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveToPosClientNode>();
    node->send_goal(76, 1);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}